nssdb/nssdb.go

```diff
@@ -0,0 +1,457 @@
+	id, err := db.getObjectID(ctx)
+	if err != nil {
+		return 0, fmt.Errorf("get new object id: %w", err)
+	}
+	obj.ID = id
+
+	if err := db.insert(ctx, obj, false); err != nil {
+		return 0, err
+	}
+	return id, nil
+}
+
+// InsertPrivate adds an object to the nssPrivate table of the key db.
+func (db *NSSDB) InsertPrivate(ctx context.Context, obj *Object) (uint32, error) {
+	id, err := db.getObjectID(ctx)
+	if err != nil {
+		return 0, fmt.Errorf("get new object id: %w", err)
+	}
+	obj.ID = id
+	obj.EncryptedAttributes = map[string][]byte{}
+	for k, plaintext := range obj.Attributes {
+		if !privateAttributes[k] {
+			continue
+		}
+		encrypted, err := db.encrypt(plaintext)
+		if err != nil {
+			return 0, fmt.Errorf("encrypt %s: %w", k, err)
+		}
+		obj.EncryptedAttributes[k] = encrypted
+
+		metadata, err := db.sign(id, plaintext)
+		if err != nil {
+			return 0, err
+		}
+		obj.Metadata = append(obj.Metadata, metadata)
+	}
+	if err := db.insert(ctx, obj, true); err != nil {
+		return 0, err
+	}
+	return id, nil
+}
+
+func (db *NSSDB) insert(ctx context.Context, obj *Object, private bool) error {
+	cols := []string{"id"}
+	vals := []any{obj.ID}
+	params := []string{"$1"}
+
+	for name, data := range obj.Attributes {
+		if _, ok := obj.EncryptedAttributes[name]; ok {
+			// insert the encrypted value, not the plaintext value
+			continue
+		}
+		col := columns[name]
+		if _, ok := db.colTable[col]; !ok {
+			return fmt.Errorf("db does not have a column for %q", name)
+		}
+		cols = append(cols, col)
+		vals = append(vals, data)
+		params = append(params, fmt.Sprintf("$%d", len(vals)))
+	}
+	for name, data := range obj.EncryptedAttributes {
+		col := columns[name]
+		if _, ok := db.colTable[col]; !ok {
+			return fmt.Errorf("db does not have a column for %q", name)
+		}
+		cols = append(cols, col)
+		vals = append(vals, data)
+		params = append(params, fmt.Sprintf("$%d", len(vals)))
+	}
+	for name, u := range obj.ULongAttributes {
+		col := columns[name]
+		if _, ok := db.colTable[col]; !ok {
+			return fmt.Errorf("db does not have a column for %q", name)
+		}
+		cols = append(cols, col)
+		vals = append(vals, encodeDBUlong(u))
+		params = append(params, fmt.Sprintf("$%d", len(vals)))
+	}
+
+	if private {
+		//nolint:gosec // trusted strings
+		q := fmt.Sprintf("INSERT INTO nssPrivate (%s) VALUES (%s)", strings.Join(cols, ", "), strings.Join(params, ", "))
+		if _, err := db.Key.ExecContext(ctx, q, vals...); err != nil {
+			return err
+		}
+	} else {
+		//nolint:gosec // trusted strings
+		q := fmt.Sprintf("INSERT INTO nssPublic (%s) VALUES (%s)", strings.Join(cols, ", "), strings.Join(params, ", "))
+		if _, err := db.Cert.ExecContext(ctx, q, vals...); err != nil {
+			return err
+		}
+	}
+
+	for _, metadata := range obj.Metadata {
+		const q = "INSERT INTO metaData (id, item1, item2) VALUES ($1, $2, $3)"
+		if _, err := db.Key.ExecContext(ctx, q, metadata.ID, metadata.Item1, metadata.Item2); err != nil {
+			return err
+		}
+	}
```

Comment:
We probably want to use a transaction for db.Key and db.Cert.

Reply:
2168975